addons/mail/static/src/model/model_field.js

```diff
@@ -792,97 +656,81 @@ class ModelField {
      *
      * @private
      * @param {mail.model} record
-     * @param {string|mail.model|<string|mail.model>[]|null} newValue
+     * @param {mail.model|mail.model[]|null} newValue
      */
     _setRelationUnlinkX2Many(record, newValue) {
-        // convert newValue to array of localId, null is considered unlink all
-        const otherLocalIds = newValue === null
-            ? [...this.read(record)]
+        // null is considered unlink all
+        const deleteOtherRecords = newValue === null
+            ? this.read(record)
             : this._setRelationConvertX2ManyValue(newValue);
-        const OtherModel = this.env.models[this.to];
+        const otherRecords = this.read(record);
 
-        const prevLocalIds = this.read(record);
         let isDeleting = false;
-        for (const otherLocalId of otherLocalIds) {
-            const wasDeleted = prevLocalIds.delete(otherLocalId);
+        for (const deleteOtherRecord of deleteOtherRecords) {
+            // unlink other record from current record
+            const wasDeleted = otherRecords.delete(deleteOtherRecord);
             if (wasDeleted) {
                 isDeleting = true;
+                // unlink current record from other records
+                deleteOtherRecord.update({
+                    [this.inverse]: [['unlink', record]],
+                });
+                // apply causality
+                if (this.isCausal) {
+                    deleteOtherRecord.delete();
+                }
             }
         }
         if (!isDeleting) {
             return;
         }
-        // unlink other record from current record
-        this.write(record, prevLocalIds);
-
-        for (const otherLocalId of otherLocalIds) {
-            const otherRecord = OtherModel.get(otherLocalId);
-            // other record may be deleted due to causality, avoid useless
-            // processing
-            if (otherRecord) {
-                // unlink current record from other record
-                otherRecord.update({
-                    [this.inverse]: [['unlink', record]],
-                });
-            }
-        }
+        // register dependents
+        this.write(record, otherRecords);
```

Comment:
This doesn't make sense, it ignores records to unlink? Also comment hints that this is super fragile design (i.e. call a function just for need of a side-effect).

Reply:
It's done above, this is due to modifying the Set in place for performance. Writing the set on itself doesn't change anything, the only effect of calling write is for rev number and dependents.

Reply:
I would prefer a dedicated method for that than rely on side-effect of `write()`. I'm not 100% convinced it's the best way, but at least I feel it's better.